components/script/dom/htmloptionelement.rs

```diff
@@ -108,51 +103,66 @@ impl HTMLOptionElement {
 
     // https://html.spec.whatwg.org/multipage/#concept-option-index
     fn index(&self) -> i32 {
-        if let Some(parent) = self.upcast::<Node>().GetParentNode() {
-            if let Some(select_parent) = parent.downcast::<HTMLSelectElement>() {
-                // return index in parent select's list of options
-                return self.index_in_select(select_parent);
-            } else if parent.is::<HTMLOptGroupElement>() {
-                if let Some(grandparent) = parent.GetParentNode() {
-                    if let Some(select_grandparent) = grandparent.downcast::<HTMLSelectElement>() {
-                        // return index in grandparent select's list of options
-                        return self.index_in_select(select_grandparent);
-                    }
-                }
+        if let Some(owner_select) = self.owner_select_element() {
+            match owner_select.list_of_options().position(|n| &*n == self) {
+                Some(index) => return index.try_into().unwrap_or(0),
+                None => {
+                    // shouldn't happen but not worth a browser panic
```

Comment:
Hrm. It might be worth saying why this shouldn't happen:

```suggestion
                    // This shouldn't happen because an option should always be in it's owner's
                    // `list_of_options`, but it isn't worth possibly panicking here if this isn't the case.
```